content/Products/OpenshiftMonitoring/collecting_metrics.md

```diff
@@ -0,0 +1,184 @@
+
+### Roll your own HTTPS server
+
+> You don't use `library-go` or don't want to run a `kube-rbac-proxy` sidecar.
+
+For instance if your component is built on top of `sigs.k8s.io/controller-runtime`, the library doesn't provide any facility to configure the HTTP authn/authz as of today and it's not even possible to configure TLS (refer to https://github.com/kubernetes-sigs/controller-runtime/issues/2073 for more details). Another case might be that your component isn't an operator.
+
+In such situations, you need to implement your own HTTPS server for `/metrics`. As explained before, it needs to require and verify the TLS client certificate using the root CA stored under the `client-ca-file` key of the `kube-system/extension-apiserver-authentication` ConfigMap.
+
+In practice, the server should:
+* Set TLSConfig's `ClientAuth` field to `RequireAndVerifyClientCert`.
+* Reload the root CA when the source ConfigMap is updated.
+* Reload the server's certificate and key when they are updated.
+
+Example: TBD
+
+## Configuring Prometheus to scrape metrics
+
+To tell Promehteus to scrape the metrics from your component, you need to create:
+* A Service object.
+* a ServiceMonitor object targeting the Service.
+
+The Service object looks like this:
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  annotations:
+    # This annotation tells the service CA operator to provision a Secret
+    # holding the certificate + key to be mounted in the pods.
+    # The Secret name is secret-kube-rbac-proxy-tls.
+    service.beta.openshift.io/serving-cert-secret-name: kube-rbac-proxy-tls
+  labels:
+    app.kubernetes.io/name: my-app
+  name: metrics
+  namespace: openshift-example
+spec:
+  ports:
+  - name: metrics
+    port: 8080
+    targetPort: metrics
+  selector:
+    app.kubernetes.io/name: my-app
+  type: ClusterIP
+```
+
+Then the ServiceMonitor object looks like:
+
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  name: my-app
+  namespace: openshift-example
+spec:
+  endpoints:
+  - interval: 30s
+    # Matches the name of the service's port.
+    port: metrics
+    scheme: https
+      # The CA file used by Prometheus to verify the server's certificate.
+      # It's the cluster's CA bundle from the service CA operator.
+      caFile: /etc/prometheus/configmaps/serving-certs-ca-bundle/service-ca.crt
+      # The name of the server (CN) in the server's certificate.
+      serverName: my-app.openshift-example.svc
+      # The client's certificate file used by Prometheus when scraping the metrics.
+      # This file is located in the Prometheus container.
+      certFile: /etc/prometheus/secrets/metrics-client-certs/tls.crt
+      # The client's key file used by Prometheus when scraping the metrics.
+      # This file is located in the Prometheus container.
+      keyFile: /etc/prometheus/secrets/metrics-client-certs/tls.key
```

Comment:
Shall we mention the field `insecureSkipVerify`? 
It allows a developer testing with the scrape target without a proper mTLS settings. Once the scrape target is working, the developer will setup the mTLS.

Reply:
I'd be wary that folks forget to turn off insecureSkipVerify.

Reply:
yeah, better be strict from the start :)